File: springboot-api/springboot-api/src/main/java/com/mehmetkaradana/springboot_api/services/AuthService.java
package com.mehmetkaradana.springboot_api.services;

import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

@Service
public class AuthService {

    public ResponseEntity<String> validateToken(HttpServletRequest request) {
        String authHeader = request.getHeader("Authorization");
        String token = (authHeader != null && authHeader.startsWith("Bearer ")) ? authHeader.substring(7) : null;

        if (token == null || token.isEmpty()) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Token is missing or not provided.");
        }


        return validateTokenWithAuthService(token);
    }

    private ResponseEntity<String> validateTokenWithAuthService(String token) {
        RestTemplate restTemplate = new RestTemplate();
        String url = "http://host.docker.internal:8084/api/auth/authenticate";

        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(token);
        HttpEntity<String> entity = new HttpEntity<>(headers);

        try {
            ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.GET, entity, String.class);

            if (response.getStatusCode().is2xxSuccessful()) {
                return ResponseEntity.ok("Valid token");
            }
            // Diğer durumlar
            return ResponseEntity.status(response.getStatusCode()).body(response.getBody());

        } catch (HttpClientErrorException e) {
            //400`s error
            if (e.getStatusCode() == HttpStatus.UNAUTHORIZED) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid token");
            } else {

                return ResponseEntity.status(e.getStatusCode()).body(e.getResponseBodyAsString());
            }
        } catch (Exception e) {
            // 5xx`s errors
            System.out.println("Error: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("An error occurred during token validation.");
        }
    }
}

File: ktor-api/src/main/kotlin/com/mehmetkaradana/services/authService.kt
package com.mehmetkaradana.services

import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.response.*

suspend fun validateToken(call: ApplicationCall): Boolean {
    val token = call.request.headers["Authorization"]?.substringAfter("Bearer ")
  //  println("Gelen Token: $token")
    if (token.isNullOrEmpty()) {
        call.respond(HttpStatusCode.Unauthorized, "Token is missing or not provided.")
        return false
    }
    return validateTokenWithAuthService(token, call)
}
suspend fun validateTokenWithAuthService(token: String, call: ApplicationCall): Boolean {
    val client = HttpClient()

    return try {
        val response: HttpResponse = client.get("http://host.docker.internal:8084/api/auth/authenticate") {
            header("Authorization", "Bearer $token")
        }
        if (response.status == HttpStatusCode.OK) {
            true // Geçerli token
        } else {
            call.respond(response.status, "invalid token: ${response.status}")
            false
        }
    } catch (e: Exception) {
        println("Error: ${e.message}")
        false
    } finally {
        client.close()
    }
}